rasa_addons/core/nlg/bftemplate.py

```diff
@@ -85,11 +89,31 @@ async def generate(
             fallback_language=fallback_language,
         )
         if "language" in message: del message["language"]
+
+        if self.__is_image_available(message, "image"):
+            message["image"] = self.__substitute_url(message, "image", message["image"])
+        if self.__is_image_available(message, "elements"):
+            #loop over the image urls in a carousel
+            for element in message["elements"]:
+                element["image_url"] = self.__substitute_url(element, "image_url", element["image_url"])
+
         metadata = message.pop("metadata", {}) or {}
         for key in metadata: message[key] = metadata[key]
 
         return message
 
+    def __is_image_available(self, message: dict, value: str):
+        """Check if an image key is available in the message dict or image_url in carousel list."""
+        return value in message.keys() and self.url_substitution_pattern
+
+    def __substitute_url(self,message:dict, key: str, url:str):
+        """Rewrite image url with the pattern found in endpoint.yaml."""
+        for item in self.url_substitution_pattern:
+            substitute = re.sub(item.get("pattern"), item.get("replacement"), message[key])
+            if substitute != url:
```

Comment:
I think it's more intuitive if **all** substitutions that can apply are applied in order, and not just the first one that can apply. Then,  the last three lines can just be `return substitute`.

Reply:
I left this part as first match replace and return, since I'm assuming every pattern of url_substitution will be used for exactly one replacement in order to avoid overriding previous replacements.